### PR TITLE
Deploy the docker image on master merge and tag push

### LINKED ISSUE #26: deploy the Docker image
Build and deploy the docker image executes gatekeeperd.

(Optional) Currently, the docker image build with [Dockerfile](https://github.com/Idein/gatekeeper/blob/master/Dockerfile) is used only for tests.
`COPY` the executable binary in `docker build` process for slim down the image for practical purposes.

### PR DESCRIPTION
**Pull request by hioki**
closes #26

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -107,3 +126,26 @@ workflows:
           requires:
             - build
 
+  deploy_latest_tag:
+    jobs:
+      - deploy:
+          matrix:
+            parameters:
+              tag_name: ["latest"]
+              target: ["arm32v7"]
```
Comment (eldesh) [inline .circleci/config.yml:135]: I hope they support ARMV6 as well for RaspberryPiZero.

Comment (hioki) [reply .circleci/config.yml]: @eldesh done by 309b34c

(I don't have a Raspberry Pi Zero, so I can't confirm if it works on a Raspberry Pi Zero.)

**Comment by hioki** [comment]
I found that it is difficult for the referring side to update automatically if only tags like `v1.2.3` are used, so I will try to leave the tags in major and minor versions.

File: .circleci/config.yml [hunk 2]
```diff
@@ -107,3 +126,26 @@ workflows:
...
+            branches:
+              only: master
+
+  deploy_pushed_tag:
+    jobs:
+      - deploy:
+          name: "deploy-$CIRCLE_TAG-<<matrix.target>>"
+          matrix:
+            parameters:
+              target: ["arm32v6", "arm32v7"]
```
Comment (eldesh) [inline .circleci/config.yml:146]: Can you deply an image for x86_64?

File: Dockerfile.arm32v6 [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# hadolint ignore=DL3015
+RUN apt-get update && \
+    apt-get install -y --install-recommends gcc-arm-linux-gnueabihf=* && \
+    rustup target add arm-unknown-linux-gnueabihf && \
+    cargo build --target=arm-unknown-linux-gnueabihf --release && \
+    cp target/arm-unknown-linux-gnueabihf/release/gatekeeperd /
+
+FROM raspbian/stretch:041518
+
+COPY --from=builder /gatekeeperd /
```
Comment (eldesh) [inline Dockerfile.arm32v6:20]: COPY --from /gatekeeperd/target/release/gatekeeperd will reduce the image size.

File: .circleci/config.yml (hunk 2 above)
Comment (hioki) [reply .circleci/config.yml]: done with f11b73e

**Comment by hioki** [comment]
> Can't we unify Dockerfile for x86_64 into a multi-stage build style as well?

done with a37c4e2

**Comment by hioki** [comment]
> I found that it is difficult for the referring side to update automatically if only tags like v1.2.3 are used, so I will try to leave the tags in major and minor versions.

done with 6bc333f

File: Dockerfile.arm32v6 (hunk 3 above)
Comment (hioki) [reply Dockerfile.arm32v6:16]: @eldesh Does this mean that the image size can be reduced by copying only the binary from the builder?
If it is, it already is.
(Maybe it's a little confusing that I'm copying the binary into the "/" on line 16.)

>     cp target/arm-unknown-linux-gnueabihf/release/gatekeeperd /

Comment (eldesh) [reply Dockerfile.arm32v6:16]: Sorry. That's exactly what I meant.

**Pull request merged**